lightning/src/ln/onion_utils.rs

```diff
@@ -167,9 +167,10 @@ pub(super) fn construct_onion_keys<T: secp256k1::Signing>(secp_ctx: &Secp256k1<T
 
 /// returns the hop data, as well as the first-hop value_msat and CLTV value we should send.
 pub(super) fn build_onion_payloads(path: &Path, total_msat: u64, mut recipient_onion: RecipientOnionFields, starting_htlc_offset: u32, keysend_preimage: &Option<PaymentPreimage>) -> Result<(Vec<msgs::OutboundOnionPayload>, u64, u32, Option<u64>), APIError> {
```

Comment:
could you please explain the changes applied to this method? from my understanding you've changed the way the bitcoin amount is computed, regardless of the payment being part of a swap. so I'm wondering why we should change this

Reply:
Effectively other than adding RGB amounts to each hop we split up the bitcoin amount into a fee part and a payment part: in non-RGB nodes the bitcoin amount is always monotonically increasing when constructing the route in reverse (notice the `.rev()` in the function): the receiver gets `x`, the previous hop gets `x + <fee to fwd to receiver>`, the previous one gets `x + <fee to fwd to next> + <fee to fwd to receiver>` and so on. So the original function could get away with only having one accumulator `cur_value_msat` that would always increase.

Now with RGB, and specifically RGB swaps, we have scenarios where this assumption doesn't hold true: for example, when we send a payment that is initially RGB and then bitcoin (i.e. the user is buying an asset) this doesn't work. At some point in the route the bitcoin amount will suddenly need to decrease, because one node in the path (the user's node) is sending more bitcoin than it's actually receiving.

So the new logic splits fees and `payment_amount`. We only accumulate fees, and for every node in the "bitcoin side" of the path we also set the `payment_amount`, while for the other nodes it will be 0.

Reply:
Just to clarify, when a "normal" payment is happening all the nodes in the route will get the same `payment_amount`, so the end result is the same as the old version.

It's just that this version can also support swaps, while the previous one couldn't.

Reply:
Thanks for clarifying, now I understood the change, which makes sense. IMO it's worth it to add a comment in the code explaining the change